lib/wallet/src/Cardano/Wallet/DB/Store/Wallets/Store.hs

```diff
@@ -21,66 +19,60 @@ module Cardano.Wallet.DB.Store.Wallets.Store
-    ( MonadThrow, throwIO )
 import Data.DBVar
     ( Store (..) )
-import Data.Delta
-    ( Base, Delta )
+import Database.Persist.Sql
+    ( SqlPersistT )
 
 import qualified Cardano.Wallet.DB.Store.Meta.Model as TxMetaStore
 
+
 mkStoreTxWalletsHistory
-    :: (Monad m, MonadThrow m)
-    => Store m DeltaTxSet
-    -> Store m DeltaTxMetaHistory
-    -> Store m DeltaTxWalletsHistory
+    :: Store (SqlPersistT IO) DeltaTxSet
+    -> QueryStore (SqlPersistT IO) QueryTxMeta DeltaTxMetaHistory
+    -> Store (SqlPersistT IO) DeltaTxWalletsHistory
 mkStoreTxWalletsHistory storeTransactions storeMeta =
-    let load = liftA2 (,)
-            <$> loadS storeTransactions
-            <*> loadS storeMeta
-        write = \(txSet,txMetaHistory) -> do
+    let load =
+            liftA2 (,)
+                <$> loadS storeTransactions
+                <*> loadS (store storeMeta)
+        write = \(txSet, txMetaHistory) -> do
             writeS storeTransactions txSet
-            writeS storeMeta txMetaHistory
+            writeS (store storeMeta) txMetaHistory
         update ma delta =
-            let (mTxSet,mWmetas) = (fst <$> ma, snd <$> ma)
+            let (mTxSet, mWmetas) = (fst <$> ma, snd <$> ma)
             in  case delta of
-            RollbackTxWalletsHistory slot -> do
-                wmetas <- loadWhenNothing mWmetas storeMeta
-                updateS storeMeta (Just wmetas)
-                    $ TxMetaStore.Rollback slot
-                let ( _metas', toBeDeletedTxSet)
-                        = TxMetaStore.rollbackTxMetaHistory slot wmetas
-                updateS storeTransactions mTxSet
-                    $ DeleteTxs toBeDeletedTxSet
-
-            ExpandTxWalletsHistory wid cs -> do
-                wmetas <- loadWhenNothing mWmetas storeMeta
-                updateS storeTransactions mTxSet
-                    $ Append
-                    $ mkTxSet
-                    $ fst <$> cs
-                updateS storeMeta (Just wmetas)
-                    $ TxMetaStore.Expand
-                    $ mkTxMetaHistory wid cs
-    in Store { loadS = load, writeS = write, updateS = update }
-
--- | Call 'loadS' from a 'Store' if the value is not already in memory.
-loadWhenNothing
-    :: (Monad m, MonadThrow m, Delta da)
-    => Maybe (Base da) -> Store m da -> m (Base da)
-loadWhenNothing (Just a) _ = pure a
-loadWhenNothing Nothing store =
-    loadS store >>= \case
-        Left (SomeException e) -> throwIO e
-        Right a -> pure a
+                    RollbackTxWalletsHistory slot -> do
+                        tbd <- case mWmetas of
+                            Nothing -> queryS storeMeta $ GetAfterSlot slot
+                            Just metas ->
+                                pure
+                                    $ snd
+                                    $ TxMetaStore.rollbackTxMetaHistory
+                                        slot
+                                        metas
```

Comment:
```suggestion
                            Just metas -> pure $ query (GetAfterSlot slot) metas
```

I'm still a bit spooked. 👻  Writing it in this way makes it obvious that the `Nothing` and the `Just` case do the same thing; whether this thing agrees with `TxMetaStore.rollbackTxMetaHistory` will then be checked by store laws.